File: ios/Packages/SolennixCore/Sources/SolennixCore/Cache/CachedEvent.swift
import Foundation
import SwiftData

// MARK: - CachedEvent

/// SwiftData model mirroring the `Event` struct for offline caching.
@Model
public final class CachedEvent {

    // MARK: - Properties

    @Attribute(.unique)
    public var id: String

    public var userId: String
    public var clientId: String
    public var eventDate: String
    public var startTime: String?
    public var endTime: String?
    public var serviceType: String
    public var numPeople: Int
    public var statusRawValue: String
    public var discount: Double
    public var discountTypeRawValue: String
    public var requiresInvoice: Bool
    public var taxRate: Double
    public var taxAmount: Double
    public var totalAmount: Double
    public var location: String?
    public var city: String?
    public var depositPercent: Double?
    public var cancellationDays: Double?
    public var refundPercent: Double?
    public var notes: String?
    public var photos: String?
    public var createdAt: String
    public var updatedAt: String

    // MARK: - Init

    public init(
        id: String,
        userId: String,
        clientId: String,
        eventDate: String,
        startTime: String? = nil,
        endTime: String? = nil,
        serviceType: String,
        numPeople: Int,
        statusRawValue: String,
        discount: Double,
        discountTypeRawValue: String,
        requiresInvoice: Bool,
        taxRate: Double,
        taxAmount: Double,
        totalAmount: Double,
        location: String? = nil,
        city: String? = nil,
        depositPercent: Double? = nil,
        cancellationDays: Double? = nil,
        refundPercent: Double? = nil,
        notes: String? = nil,
        photos: String? = nil,
        createdAt: String,
        updatedAt: String
    ) {
        self.id = id
        self.userId = userId
        self.clientId = clientId
        self.eventDate = eventDate
        self.startTime = startTime
        self.endTime = endTime
        self.serviceType = serviceType
        self.numPeople = numPeople
        self.statusRawValue = statusRawValue
        self.discount = discount
        self.discountTypeRawValue = discountTypeRawValue
        self.requiresInvoice = requiresInvoice
        self.taxRate = taxRate
        self.taxAmount = taxAmount
        self.totalAmount = totalAmount
        self.location = location
        self.city = city
        self.depositPercent = depositPercent
        self.cancellationDays = cancellationDays
        self.refundPercent = refundPercent
        self.notes = notes
        self.photos = photos
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }

    /// Creates a cached version from an `Event` struct.
    public convenience init(from event: Event) {
        self.init(
            id: event.id,
            userId: event.userId,
            clientId: event.clientId,
            eventDate: event.eventDate,
            startTime: event.startTime,
            endTime: event.endTime,
            serviceType: event.serviceType,
            numPeople: event.numPeople,
            statusRawValue: event.status.rawValue,
            discount: event.discount,
            discountTypeRawValue: event.discountType.rawValue,
            requiresInvoice: event.requiresInvoice,
            taxRate: event.taxRate,
            taxAmount: event.taxAmount,
            totalAmount: event.totalAmount,
            location: event.location,
            city: event.city,
            depositPercent: event.depositPercent,
            cancellationDays: event.cancellationDays,
            refundPercent: event.refundPercent,
            notes: event.notes,
            photos: event.photos,
            createdAt: event.createdAt,
            updatedAt: event.updatedAt
        )
    }

    // MARK: - Conversion

    /// Converts this cached model back to an `Event` value type.
    public func toEvent() -> Event {
        Event(
            id: id,
            userId: userId,
            clientId: clientId,
            eventDate: eventDate,
            startTime: startTime,
            endTime: endTime,
            serviceType: serviceType,
            numPeople: numPeople,
            status: EventStatus(rawValue: statusRawValue) ?? .quoted,
            discount: discount,
            discountType: DiscountType(rawValue: discountTypeRawValue) ?? .percent,
            requiresInvoice: requiresInvoice,
            taxRate: taxRate,
            taxAmount: taxAmount,
            totalAmount: totalAmount,
            location: location,
            city: city,
            depositPercent: depositPercent,
            cancellationDays: cancellationDays,
            refundPercent: refundPercent,
            notes: notes,
            photos: photos,
            createdAt: createdAt,
            updatedAt: updatedAt
        )
    }
}

File: android/core/database/src/main/java/com/creapolis/solennix/core/database/entity/CachedEvent.kt
package com.creapolis.solennix.core.database.entity

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import com.creapolis.solennix.core.model.DiscountType
import com.creapolis.solennix.core.model.Event
import com.creapolis.solennix.core.model.EventStatus

enum class SyncStatus {
    SYNCED, PENDING_INSERT, PENDING_UPDATE, PENDING_DELETE
}

@Entity(tableName = "events")
data class CachedEvent(
    @PrimaryKey val id: String,
    @ColumnInfo(name = "user_id") val userId: String,
    @ColumnInfo(name = "client_id") val clientId: String,
    @ColumnInfo(name = "event_date") val eventDate: String,
    @ColumnInfo(name = "start_time") val startTime: String?,
    @ColumnInfo(name = "end_time") val endTime: String?,
    @ColumnInfo(name = "service_type") val serviceType: String,
    @ColumnInfo(name = "num_people") val numPeople: Int,
    val status: EventStatus,
    val discount: Double,
    @ColumnInfo(name = "discount_type") val discountType: DiscountType,
    @ColumnInfo(name = "requires_invoice") val requiresInvoice: Boolean,
    @ColumnInfo(name = "tax_rate") val taxRate: Double,
    @ColumnInfo(name = "tax_amount") val taxAmount: Double,
    @ColumnInfo(name = "total_amount") val totalAmount: Double,
    val location: String?,
    val city: String?,
    @ColumnInfo(name = "deposit_percent") val depositPercent: Double?,
    @ColumnInfo(name = "cancellation_days") val cancellationDays: Double?,
    @ColumnInfo(name = "refund_percent") val refundPercent: Double?,
    val notes: String?,
    val photos: String?,
    @ColumnInfo(name = "created_at") val createdAt: String,
    @ColumnInfo(name = "updated_at") val updatedAt: String,
    @ColumnInfo(name = "sync_status") val syncStatus: SyncStatus = SyncStatus.SYNCED
)

fun CachedEvent.asExternalModel() = Event(
    id = id,
    userId = userId,
    clientId = clientId,
    eventDate = eventDate,
    startTime = startTime,
    endTime = endTime,
    serviceType = serviceType,
    numPeople = numPeople,
    status = status,
    discount = discount,
    discountType = discountType,
    requiresInvoice = requiresInvoice,
    taxRate = taxRate,
    taxAmount = taxAmount,
    totalAmount = totalAmount,
    location = location,
    city = city,
    depositPercent = depositPercent,
    cancellationDays = cancellationDays,
    refundPercent = refundPercent,
    notes = notes,
    photos = photos,
    createdAt = createdAt,
    updatedAt = updatedAt
)

fun Event.asEntity(syncStatus: SyncStatus = SyncStatus.SYNCED) = CachedEvent(
    id = id,
    userId = userId,
    clientId = clientId,
    eventDate = eventDate,
    startTime = startTime,
    endTime = endTime,
    serviceType = serviceType,
    numPeople = numPeople,
    status = status,
    discount = discount,
    discountType = discountType,
    requiresInvoice = requiresInvoice,
    taxRate = taxRate,
    taxAmount = taxAmount,
    totalAmount = totalAmount,
    location = location,
    city = city,
    depositPercent = depositPercent,
    cancellationDays = cancellationDays,
    refundPercent = refundPercent,
    notes = notes,
    photos = photos,
    createdAt = createdAt,
    updatedAt = updatedAt,
    syncStatus = syncStatus
)
